renku/core/commands/update.py

```diff
@@ -46,96 +52,211 @@ def update_workflows():
 
-    wf, path = CWLConverter.convert(workflow, client.path)
-    # Don't compute paths if storage is disabled.
+    # NOTE: Pull inputs from Git LFS or other storage backends
     if client.check_external_storage():
-        # Make sure all inputs are pulled from a storage.
-        paths_ = (i.consumes.path for i in workflow.inputs)
-        client.pull_paths_from_storage(*paths_)
+        inputs = [i.actual_value for p in plans for i in p.inputs]
+        client.pull_paths_from_storage(*inputs)
 
     delete_indirect_files_list(client.path)
 
-    # Execute the workflow and relocate all output files.
-    # FIXME get new output paths for edited tools
-    # output_paths = {path for _, path in workflow.iter_output_files()}
-    execute(path, output_paths=output_paths)
+    started_at_time = local_now()
 
-    paths = [o.produces.path for o in workflow.outputs]
+    modified_outputs = _execute_workflow_helper(plans=plans, client=client)
 
-    add_to_git(client.repo.git, *paths)
+    ended_at_time = local_now()
 
-    if client.repo.is_dirty():
-        commit_msg = f"renku {command_name}: committing {len(paths)} newly added files"
+    add_to_git(client.repo.git, *modified_outputs)
 
+    if client.repo.is_dirty():
+        postfix = "s" if len(modified_outputs) > 1 else ""
+        commit_msg = f"renku {command_name}: committing {len(modified_outputs)} modified file{postfix}"
         committer = Actor(f"renku {__version__}", version_url)
-
         client.repo.index.commit(commit_msg, committer=committer, skip_hooks=True)
 
-    workflow_name = f"{uuid.uuid4().hex}_{command_name}.yaml"
-
-    path = client.workflow_path / workflow_name
-
-    workflow.update_id_and_label_from_commit_path(client, client.repo.head.commit, path)
-
-    # TODO: implement properly with new database
-    # if not workflow.subprocesses:  # Update parameters if there is only one step
-    #     _update_run_parameters(run=workflow, working_dir=client.path)
-
-    # cls = WorkflowRun if workflow.subprocesses else ProcessRun
-    # activity = cls.from_run(run=workflow, path=path, update_commits=update_commits)
-    # activity.to_yaml(path=path)
-    # client.add_to_activity_index(activity)
-
-    # activity_gateway.add(activity)
-
-
-# def _update_run_parameters(run, working_dir):
-
-#     default_parameters = {p.name: p for p in run.run_parameters}
-
-#     indirect_parameters = read_indirect_parameters(working_dir)
-# for name, value in indirect_parameters.items():
-#     id_ = RunParameter.generate_id(run_id=run._id, name=name)
-#     parameter = RunParameter(id=id_, name=name, value=value)
-#     default_parameters[name] = parameter
-
-# run.run_parameters = list(default_parameters.values())
+    database = database_dispatcher.current_database
+
+    activities = []
+
+    for plan in plans:
+        # NOTE: Update plans are copies of Plan objects. We need to use the original Plan objects to avoid duplicates.
+        original_plan = database["plans"].get(plan.id)
```

Comment:
We shouldn't access the database directly and instead go through the interfaces, to make sure the code is properly insulated from implementation details of the database. So `IPlanGateway.get_by_id(plan.id)`